### ISSUE TITLE
How do I get validation loss with custom dataset?

### ISSUE THREAD

**Issue by frandrieu** [open-mmlab/mmdetection#8332]
Hello,

First of all. thank you for this excellent framework. I work on an object detector with faster_rcnn on a custom dataset, and I would like to get the validation loss, but it seems that the only validation metric I can get are AP50 or recall... Did I miss something? Is there a way around that? 

Thank you for your help

**Comment by BIGWangYuDong** [comment]
Sorry for the late reply, I think you can follow https://github.com/open-mmlab/mmdetection/issues/7971

**Comment by frandrieu** [comment]
Hi, 

I partly solved my problem using [this issue](https://github.com/open-mmlab/mmsegmentation/issues/483): 



```
from mmdet.datasets import build_dataset
from mmdet.models import build_detector
from mmdet.apis import train_detector
from mmdet.utils import get_device

datasets = [build_dataset(cfg.data.train)]

if len(cfg.workflow) == 2:
  val_dataset = copy.deepcopy(cfg.data.val)
  val_dataset.pipeline = cfg.data.train.pipeline
  datasets.append(build_dataset(val_dataset))

model = build_detector(cfg.model)
model.CLASSES = datasets[0].CLASSES
mmcv.mkdir_or_exist(osp.abspath(cfg.work_dir))

train_detector(model, datasets, cfg, distributed=False, validate=True)
```


Then validation loss I get seemed to overwrite the training loss. Apparently, the training loss and validation loss are retuned with the same key. I have also found [this issue](https://github.com/open-mmlab/mmsegmentation/issues/1396) about this, so I replaced the 'val_step function' in base.py to add the new keys for validation:

```
    def val_step(self, data_batch, optimizer=None, **kwargs):
        """The iteration step during validation.

        This method shares the same signature as :func:`train_step`, but used
        during val epochs. Note that the evaluation after training epochs is
        not implemented with this method, but an evaluation hook.
        """
        losses = self(**data_batch)
        loss, log_vars = self._parse_losses(losses)
        log_vars_ = dict()
        for loss_name, loss_value in log_vars.items():
            k = loss_name + '_val'
            log_vars_[k] = loss_value

        outputs = dict(
            loss=loss,
            log_vars=log_vars_,
            num_samples=len(data_batch['img_metas']))

        return outputs
```

 
Now I get all the losses (training and validation) into the "val" tab when I visualize the log with tensorboard... Any suggestion?

**Comment by BIGWangYuDong** [comment]
I think there is a parameter named 'training' in the detector, you may add a condition, such as
```
if self.training:
    xxx
```
as for tensorboard, I did not get your point

**Comment by BIGWangYuDong** [comment]
feel free to reopen this issue if you still have problems

**Comment by MARUIHAN123** [comment]
@frandrieu  Where we can modify this file code
